sdk/eventhub/azure-eventhubs/azure/eventhub/_client_base.py

```diff
@@ -0,0 +1,373 @@
+        else:
+            host = address
+        return cls(host, entity, EventHubSharedKeyCredential(policy, key), **kwargs)
+
+    def _create_auth(self):
+        """
+        Create an ~uamqp.authentication.SASTokenAuth instance to authenticate
+        the session.
+        """
+        http_proxy = self._config.http_proxy
+        transport_type = self._config.transport_type
+        auth_timeout = self._config.auth_timeout
+
+        # TODO: the following code can be refactored to create auth from classes directly instead of using if-else
+        if isinstance(self._credential, EventHubSharedKeyCredential):  # pylint:disable=no-else-return
+            username = self._credential.policy
+            password = self._credential.key
+            if "@sas.root" in username:
+                return authentication.SASLPlain(
+                    self._address.hostname, username, password, http_proxy=http_proxy, transport_type=transport_type)
+            return authentication.SASTokenAuth.from_shared_access_key(
+                self._auth_uri, username, password, timeout=auth_timeout, http_proxy=http_proxy,
+                transport_type=transport_type)
+
+        elif isinstance(self._credential, EventHubSASTokenCredential):
+            token = self._credential.get_sas_token()
+            try:
+                expiry = int(parse_sas_token(token)['se'])
+            except (KeyError, TypeError, IndexError):
+                raise ValueError("Supplied SAS token has no valid expiry value.")
+            return authentication.SASTokenAuth(
+                self._auth_uri, self._auth_uri, token,
+                expires_at=expiry,
+                timeout=auth_timeout,
+                http_proxy=http_proxy,
+                transport_type=transport_type)
+
+        else:  # Azure credential
+            get_jwt_token = functools.partial(self._credential.get_token, JWT_TOKEN_SCOPE)
+            return authentication.JWTTokenAuth(self._auth_uri, self._auth_uri,
+                                               get_jwt_token, http_proxy=http_proxy,
+                                               transport_type=transport_type)
+
+    def _close_connection(self):
+        self._conn_manager.reset_connection_if_broken()
+
+    def _try_delay(self, retried_times, last_exception, timeout_time=None, entity_name=None):
+        entity_name = entity_name or self._container_id
+        backoff = self._config.backoff_factor * 2 ** retried_times
+        if backoff <= self._config.backoff_max and (
+                timeout_time is None or time.time() + backoff <= timeout_time):  # pylint:disable=no-else-return
+            time.sleep(backoff)
+            _LOGGER.info("%r has an exception (%r). Retrying...", format(entity_name), last_exception)
+        else:
+            _LOGGER.info("%r operation has timed out. Last exception before timeout is (%r)",
+                     entity_name, last_exception)
+            raise last_exception
```

Comment:
Same statement as for our other "try_<foo>":  If we're raising/not returning a bool, it's not really a try, it's just like, "_perform_backoff_delay" (This is verbose, but that's my nature, so feel free to rejigger)

Reply:
Fixed!